packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewFlutterPlugin.java

```diff
@@ -38,36 +50,100 @@ public WebViewFlutterPlugin() {}
     new FlutterCookieManager(registrar.messenger());
   }
 
+  private static void setUp(
+      BinaryMessenger binaryMessenger,
+      PlatformViewRegistry viewRegistry,
+      Activity activity,
+      View containerView) {
+    InstanceManager instanceManager = new InstanceManager();
+    viewRegistry.registerViewFactory(
+        "plugins.flutter.io/webview", new FlutterWebViewFactory(instanceManager));
+    WebViewHostApi.setup(
+        binaryMessenger,
+        new WebViewHostApiImpl(
+            instanceManager, new WebViewHostApiImpl.WebViewProxy(), activity, containerView));
+    WebViewClientHostApi.setup(
+        binaryMessenger,
+        new WebViewClientHostApiImpl(
+            instanceManager,
+            new WebViewClientHostApiImpl.WebViewClientCreator(),
+            new WebViewClientFlutterApiImpl(binaryMessenger, instanceManager)));
+    WebChromeClientHostApi.setup(
+        binaryMessenger,
+        new WebChromeClientHostApiImpl(
+            instanceManager,
+            new WebChromeClientHostApiImpl.WebChromeClientCreator(),
+            new WebChromeClientFlutterApiImpl(binaryMessenger, instanceManager)));
+    DownloadListenerHostApi.setup(
+        binaryMessenger,
+        new DownloadListenerHostApiImpl(
+            instanceManager,
+            new DownloadListenerHostApiImpl.DownloadListenerCreator(),
+            new DownloadListenerFlutterApiImpl(binaryMessenger, instanceManager)));
+    JavaScriptChannelHostApi.setup(
+        binaryMessenger,
+        new JavaScriptChannelHostApiImpl(
+            instanceManager,
+            new JavaScriptChannelHostApiImpl.JavaScriptChannelCreator(),
+            new JavaScriptChannelFlutterApiImpl(binaryMessenger, instanceManager),
+            new Handler(activity.getMainLooper())));
+    WebSettingsHostApi.setup(
+        binaryMessenger,
+        new WebSettingsHostApiImpl(
+            instanceManager, new WebSettingsHostApiImpl.WebSettingsCreator()));
+    new FlutterCookieManager(binaryMessenger);
+  }
+
   @Override
   public void onAttachedToEngine(FlutterPluginBinding binding) {
-    BinaryMessenger messenger = binding.getBinaryMessenger();
-    binding
-        .getPlatformViewRegistry()
-        .registerViewFactory(
-            "plugins.flutter.io/webview",
-            new FlutterWebViewFactory(messenger, /*containerView=*/ null));
-    flutterCookieManager = new FlutterCookieManager(messenger);
+    this.pluginBinding = binding;
+    flutterCookieManager = new FlutterCookieManager(binding.getBinaryMessenger());
   }
 
   @Override
-  public void onDetachedFromEngine(FlutterPluginBinding binding) {
+  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
     if (flutterCookieManager == null) {
       return;
     }
 
     flutterCookieManager.dispose();
     flutterCookieManager = null;
   }
+
+  @Override
+  public void onAttachedToActivity(@NonNull ActivityPluginBinding activityPluginBinding) {
+    setUp(
```

Comment:
Can't `onAttachedToActivity` be called multiple times? What are the implications of moving platform view registration from the engine attachment to the activity attachment?

And similarly, are the pigeon setup APIs safe to call repeatedly?

Reply:
We had a similar problem with doing this with [google_mobile_ads](https://github.com/googleads/googleads-mobile-flutter/pull/381). It mostly only affect loading views with cached engines. The solution was to set everything up in `onAttachedToEngine` with `binding.applicationContext` and then switch the `context` when the plugin was attached to an `Activity`. I went ahead and changed this plugin to do the same. I don't believe there are any other problems this implementation causes.